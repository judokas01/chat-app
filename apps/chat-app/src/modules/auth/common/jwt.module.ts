import { JwtModule } from '@nestjs/jwt'
import { ConfigService } from '@root/common/config/config-service.service'

export const JWT = JwtModule.registerAsync({
    extraProviders: [ConfigService],
    global: true,
    inject: [ConfigService],
    // eslint-disable-next-line require-await
    useFactory: async (configService: ConfigService) => {
        const { expiresIn, secret } = configService.config.auth.jwt
        return {
            secretOrPrivateKey: secret,
            signOptions: {
                expiresIn: expiresIn,
            },
        }
    },
})
