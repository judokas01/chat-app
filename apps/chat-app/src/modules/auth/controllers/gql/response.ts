import { Field, ObjectType } from '@nestjs/graphql'

@ObjectType()
export class LoginResponse {
    @Field()
    accessToken: string

    @Field()
    renewToken: string
}

@ObjectType()
export class RegisterResponse {
    @Field()
    id: string

    @Field()
    userName: string

    @Field()
    email: string
}
