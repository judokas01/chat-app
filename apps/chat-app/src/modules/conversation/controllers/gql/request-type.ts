import { ArgsType, Field } from '@nestjs/graphql'

@ArgsType()
export class GetConversationArgsGql {
    @Field()
    userId: string
}

@ArgsType()
export class FindUsersArgsGql {
    @Field({ nullable: true })
    userName: string

    @Field({ nullable: true })
    email: string
}

@ArgsType()
export class CreateConversationArgsGql {
    @Field(() => [String])
    userIds: string[]

    @Field({ nullable: true })
    name: string
}
