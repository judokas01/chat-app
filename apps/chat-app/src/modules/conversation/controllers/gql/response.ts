import { Field, ObjectType } from '@nestjs/graphql'

@ObjectType()
export class Conversation {
    @Field(() => String)
    id: string

    @Field(() => String, { nullable: true })
    name: string | null

    @Field(() => Date)
    createdAt: Date

    @Field(() => Date, { nullable: true })
    lastMessageAt: Date | null

    @Field(() => [ConversationUser], { nullable: true })
    users: ConversationUser[]
}

@ObjectType()
export class ConversationUser {
    @Field(() => String)
    id: string

    @Field(() => String)
    userName: string

    @Field(() => String)
    email: string
}

@ObjectType()
export class ConversationMessageResponse {
    @Field(() => String)
    cursor: string

    @Field(() => Boolean)
    hasMore: boolean

    @Field(() => [Message])
    messages: Message[]
}

@ObjectType()
export class Message {
    @Field(() => String)
    id: string

    @Field(() => String)
    text: string

    @Field(() => ConversationUser)
    author: ConversationUser

    @Field(() => Date)
    createdAt: Date

    @Field(() => Boolean)
    isRemoved: boolean
}
