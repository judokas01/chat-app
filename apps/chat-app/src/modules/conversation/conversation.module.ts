import { Module, ValidationPipe } from '@nestjs/common'
import { ConfigService } from '@root/common/config/config-service.service'
import { PrismaService } from '@root/infrastructure/prisma/prisma.service'
import { IUserRepository } from '@root/common/repositories/user.repository'
import { UserPrismaRepository } from '@root/common/repositories/user/prisma/user.repository'
import { IMessageRepository } from '@root/common/repositories/message.repository'
import { MessagePrismaRepository } from '@root/common/repositories/message/prisma/message.repository'
import { JwtService } from '@nestjs/jwt'
import { AuthGuard, IAuthGuard } from '@root/common/guards/authenticate/authenticate.guard'
import { JwtAuthGuard } from '@root/common/guards/authenticate/services/jwt-authenticate.service'
import { IConversationRepository } from '../../common/repositories/conversation.repository'
import { PrismaConversationRepository } from '../../common/repositories/conversation/prisma/conversation.repository'
import { CreateConversationService } from './create-conversation/create-conversation.service'
import { MessageService } from './message/message.service'
import { ConversationController } from './controllers/rest/conversation.controller'
import { ConversationResolver } from './controllers/gql/conversation.resolver'
import { GetConversationService } from './get-conversation/get-conversation.service'
import { FindUserService } from './find-user/find-user.service'

@Module({
    controllers: [ConversationController],
    exports: [
        { provide: IConversationRepository, useClass: PrismaConversationRepository },
        { provide: IUserRepository, useClass: UserPrismaRepository },
        { provide: IMessageRepository, useClass: MessagePrismaRepository },
        ConfigService,
        PrismaService,
        CreateConversationService,
        MessageService,
        ConversationResolver,
        GetConversationService,
        FindUserService,
        ValidationPipe,
        JwtService,
    ],
    providers: [
        { provide: IConversationRepository, useClass: PrismaConversationRepository },
        { provide: IUserRepository, useClass: UserPrismaRepository },
        { provide: IMessageRepository, useClass: MessagePrismaRepository },
        { provide: IAuthGuard, useClass: JwtAuthGuard },
        AuthGuard,
        ConfigService,
        PrismaService,
        CreateConversationService,
        MessageService,
        ConversationResolver,
        GetConversationService,
        FindUserService,
        ValidationPipe,
        JwtService,
    ],
})
export class ConversationModule {}
