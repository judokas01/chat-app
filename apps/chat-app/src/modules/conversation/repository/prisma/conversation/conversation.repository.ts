import { Injectable } from '@nestjs/common'
import { PrismaService } from '@root/infrastructure/prisma/prisma.service'
import { IConversationRepository } from '../../IConversationRepository'

@Injectable()
export class PrismaConversationRepository implements IConversationRepository {
    constructor(private prisma: PrismaService) {}
}
