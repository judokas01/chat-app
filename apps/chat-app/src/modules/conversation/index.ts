import { Module } from '@nestjs/common'
import { ConfigService } from '@root/common/config/config-service.service'
import { PrismaService } from '@root/infrastructure/prisma/prisma.service'
import { IConversationRepository } from '../../common/repositories/conversation.repository'
import { PrismaConversationRepository } from '../../common/repositories/conversation/prisma/conversation/conversation.repository'

@Module({
    providers: [
        { provide: IConversationRepository, useClass: PrismaConversationRepository },
        ConfigService,
        PrismaService,
    ],
})
export class ConversationModule {}
