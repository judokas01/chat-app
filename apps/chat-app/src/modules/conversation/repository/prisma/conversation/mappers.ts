import { Conversation, ConversationInput } from '@root/common/entities/conversation.entity'
import {
    Prisma,
    UsersConversation as PrismaUserConversation,
    Message as PrismaMessage,
    Conversation as PrismaConversation,
} from '@prisma/client'

import { HasMany } from '@root/common/entities/common/Relationship'

export const toCoreConversation = (
    conversation: PrismaConversation & {
        messages?: PrismaMessage[]
        usersConversation?: PrismaUserConversation[]
    },
): Conversation => {
    const { id, createdAt, lastMessageAt, messages, usersConversation, customName } = conversation

    return {
        createdAt,
        id,
        lastMessageAt,
        messages: toCoreMessages(messages),
        name: customName,
        participants: toCoreParticipant(usersConversation),
    }
}

const toCoreParticipant = (
    userConversation?: PrismaUserConversation[],
): Conversation['participants'] => {
    if (!userConversation) {
        return new HasMany(undefined, 'conversation.participants')
    }

    return new HasMany(
        userConversation.map(({ userId }) => userId),
        'conversation.participants',
    )
}

const toCoreMessages = (messages?: PrismaMessage[]): Conversation['messages'] => {
    if (!messages) {
        return new HasMany(undefined, 'conversation.messages')
    }
    return new HasMany(
        messages.map(({ id }) => id),
        'conversation.participants',
    )
}

export const toConversationCreate = (
    conversation: ConversationInput,
): Prisma.ConversationCreateInput => {
    const { name } = conversation

    return {
        customName: name,
        lastMessageAt: null,
    }
}
