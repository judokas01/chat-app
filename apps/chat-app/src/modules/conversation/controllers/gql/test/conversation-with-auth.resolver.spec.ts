import { describe, it, expect, beforeEach, afterEach } from 'vitest'
import {
    getTestModuleWithInterface,
    TestInterfaceModule,
} from '@root/common/test-utilities/test-app/interface'
import { userMock } from '@root/common/test-utilities/mocks/user'
import { conversationMock } from '@root/common/test-utilities/mocks/conversation'
import { faker } from '@faker-js/faker'
import { User } from '@root/common/entities/user.entity'
import { IAuthGuard } from '@root/common/guards/authenticate/authenticate.guard'
import { JwtAuthGuard } from '@root/common/guards/authenticate/services/jwt-authenticate.service'
import { JWTPayload } from '@root/modules/auth/common/types'
import { JwtService } from '@nestjs/jwt'
import { ConversationModule } from '../../../conversation.module'
import { ConversationUser, Conversation as GqlConversation } from '../response'
import { FindUsersArgsGql } from '../request-type'
import { findUsersGqlRequest, getUserConversationGqlRequest } from './helpers'

describe('ConversationResolver - guard tests', () => {
    let testModule: TestInterfaceModule
    let token: string
    let authUser: User

    beforeEach(async () => {
        testModule = await getTestModuleWithInterface({
            module: ConversationModule,
            providers: [{ provide: IAuthGuard, useClass: JwtAuthGuard }],
        })

        await testModule.cleanDb()

        const service = testModule.module.get<JwtService>(JwtService)
        authUser = await userMock.random.createOne({}, testModule)
        const jwtPayload: JWTPayload = { sub: authUser.id, userName: authUser.data.userName }
        token = `Bearer ` + (await service.signAsync(jwtPayload))
    })

    afterEach(async () => {
        await testModule.destroy()
    })

    it('should throw unauthorized error when invalid token is provided', async () => {
        const numberOfConversations = faker.number.int({ max: 5, min: 1 })
        const user = await userMock.random.createOne({}, testModule)

        await Promise.all(
            Array.from({ length: numberOfConversations }).map(() =>
                conversationMock.createOne({ user }, testModule),
            ),
        )

        const conversations = await testModule.requestGql
            .send(getUserConversationGqlRequest({ userId: user.id }))
            .set('Authorization', 'invalidToken')

        expect(conversations.body.errors).toMatchSnapshot()
    })

    it('should throw unauthorized error, when auth user is not the same as userId requested', async () => {
        const numberOfConversations = faker.number.int({ max: 5, min: 1 })
        const user = await userMock.random.createOne({}, testModule)

        await Promise.all(
            Array.from({ length: numberOfConversations }).map(() =>
                conversationMock.createOne({ user }, testModule),
            ),
        )

        const conversations = await testModule.requestGql
            .send(getUserConversationGqlRequest({ userId: user.id }))
            .set('Authorization', token)

        expect(conversations.body.errors).toMatchSnapshot()
    })

    it('should authenticate user when auth user matches the requested userId', async () => {
        const numberOfConversations = faker.number.int({ max: 5, min: 1 })

        await Promise.all(
            Array.from({ length: numberOfConversations }).map(() =>
                conversationMock.createOne({ user: authUser }, testModule),
            ),
        )

        const conversations = await testModule.requestGql
            .send(getUserConversationGqlRequest({ userId: authUser.id }))
            .set('Authorization', token)

        expect(conversations.body.errors).toBeUndefined()
        expect(conversations.body.data.getUserConversations.at(0)).toMatchObject({
            createdAt: expect.any(String),
            id: expect.any(String),
            lastMessageAt: null,
            name: null,
            users: expect.any(Array),
        } satisfies GqlConversation)
    })

    it.each([
        {
            getArgs: (user: User): FindUsersArgsGql => ({ email: user.data.email }),
            text: 'by email',
        },
        {
            getArgs: (user: User): FindUsersArgsGql => ({ userName: user.data.userName }),
            text: 'by userName',
        },
        {
            getArgs: (user: User): FindUsersArgsGql => ({
                email: user.data.email,
                userName: user.data.userName,
            }),
            text: 'by both',
        },
    ])('should return all users found by $text', async ({ getArgs }) => {
        const user = await userMock.random.createOne({}, testModule)

        const conversations = await testModule.requestGql.send(findUsersGqlRequest(getArgs(user)))

        const items = conversations.body.data.findUsers as ConversationUser[]
        expect(items).toHaveLength(1)

        items.forEach((conversation) => {
            expect(conversation).toMatchObject({
                email: user.data.email,
                id: user.id,
                userName: user.data.userName,
            } satisfies ConversationUser)
        })
    })
})
