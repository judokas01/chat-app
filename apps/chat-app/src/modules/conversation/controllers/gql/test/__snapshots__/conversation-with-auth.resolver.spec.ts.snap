// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConversationResolver - guard tests > should throw unauthorized error when invalid token is provided 1`] = `
[
  {
    "extensions": {
      "code": "UNAUTHENTICATED",
      "originalError": {
        "message": "Unauthorized",
        "statusCode": 401,
      },
    },
    "locations": [
      {
        "column": 13,
        "line": 3,
      },
    ],
    "message": "Unauthorized",
    "path": [
      "getUserConversations",
    ],
  },
]
`;

exports[`ConversationResolver - guard tests > should throw unauthorized error, when auth user is not the same as userId requested 1`] = `
[
  {
    "extensions": {
      "code": "UNAUTHENTICATED",
      "originalError": {
        "error": "Unauthorized",
        "message": "Cannot get conversations of another user",
        "statusCode": 401,
      },
    },
    "locations": [
      {
        "column": 13,
        "line": 3,
      },
    ],
    "message": "Cannot get conversations of another user",
    "path": [
      "getUserConversations",
    ],
  },
]
`;
